### PR TITLE
Union bugfixes

### LINKED ISSUE #1046: UNION queries entities should be independent
UNION queries entities should be independent
```
MATCH (n)-[t]->() WHERE ID(n) IN [1, 3, 5] RETURN t
UNION
MATCH ()-[t]->(m) WHERE ID (m) IN [1, 3, 5] RETURN t

(error) Cannot use the same relationship variable 't' for multiple patterns.
```

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Resolves #1046, fixing  the incorrect syntax error:
```
MATCH ()-[e]->() RETURN e UNION MATCH ()-[e]->() RETURN e"
(error) Cannot use the same relationship variable 'e' for multiple patterns.
```

Fixes a bug in OpDistinct logic on columns containing both nodes and edges (also, adds our first test for such columns!).

### CODE REVIEW

File: tests/flow/test_union.py [hunk 1]
```diff
@@ -41,3 +64,58 @@ def test_invalid_union(self):
...
+        union_query = """MATCH (a)-[]->(b) RETURN a.v, b.v ORDER BY a.v, b.v
+                         UNION
+                         MATCH (a)-[]->(b) RETURN a.v, b.v ORDER BY a.v, b.v"""
+        union_result = redis_graph.query(union_query)
+        self.env.assertEquals(union_result.result_set, non_union_result.result_set)
+
+    # A syntax error should be raised on edge alias reuse in one side of a union.
+    def test04_union_invalid_reused_edge(self):
+        try:
+            query = """MATCH ()-[e]->()-[e]->() RETURN e
```
Comment (swilly22) [inline tests/flow/test_union.py:83]: An error should be (and is being omitted) regardless if we're dealing with a UNION query,
```
GRAPH.EXPLAIN g "MATCH ()-[e]->()-[e]->() RETURN e"
(error) Cannot use the same relationship variable 'e' for multiple patterns.
```

File: tests/flow/test_union.py [hunk 2]
```diff
@@ -12,27 +12,50 @@ def __init__(self):
...
+        # (v1)-[:E1]->(v2)-[:E2]->(v3)
+        node_props = ['v1', 'v2', 'v3']
+
+        nodes = {}
+        for idx, v in enumerate(node_props):
+            node = Node(label="L", properties={"v": v})
+            nodes[v] = node
+            redis_graph.add_node(node)
+
+        edge = Edge(nodes['v1'], "E1", nodes['v2'], properties={"v": True})
```
Comment (swilly22) [inline tests/flow/test_union.py:29]: It would be clearer if `v` be set to 'v1_v2'

File: tests/flow/test_union.py [hunk 3]
```diff
@@ -41,3 +64,58 @@ def test_invalid_union(self):
...
+
+        expected_result = [[False],
+                           [True]]
+        self.env.assertEquals(result.result_set, expected_result)
+
+    # Union should be capable of collating nodes and edges in a single column.
+    def test06_union_nodes_with_edges(self):
+        query = """MATCH ()-[e]->() RETURN e
+                   UNION
+                   MATCH (e) RETURN e"""
```
Comment (swilly22) [inline tests/flow/test_union.py:109]: Interesting!

File: src/ast/ast_validations.c [hunk 4]
```diff
@@ -1638,6 +1628,40 @@ static AST_Validation _AST_Validate_ParseResultRoot(RedisModuleCtx *ctx,
...
+		AST_Free(scoped_ast);
+		if(res != AST_VALID) goto cleanup;
+
+		// Update the starting index of the scope for the next iteration..
+		scope_start = scope_end;
+	}
+
+	// Build and test the final scope (from the last UNION to the last clause)
+	scope_end = cypher_ast_query_nclauses(mock_ast->root);
+	scoped_ast = _NewMockASTSegment(mock_ast->root, scope_start, scope_end);
```
Comment (swilly22) [inline src/ast/ast_validations.c:1656]: How about appending `cypher_ast_query_nclauses(mock_ast->root)` to `query_scopes` that way the for loop can handle the last clause.

File: tests/flow/test_union.py (hunk 1 above)
Comment (jeffreylovitz) [reply tests/flow/test_union.py:83]: Correct. This test is to prove that standard AST validations still work when we follow the `_AST_ValidateUnionQuery` code path.

File: src/ast/ast_validations.c (hunk 4 above)
Comment (jeffreylovitz) [reply src/ast/ast_validations.c:1656]: Good idea, thanks!

**Pull request merged**